src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.apache.spark.api.dotnet
+
+import java.util.concurrent.{ExecutorService, Executors}
+
+import scala.collection.mutable
+
+/**
+ * Pool of thread executors. There should be a 1-1 correspondence between C# threads
+ * and Java threads.
+ */
+object ThreadPool {
+
+  /**
+   * Map from threadId to corresponding executor.
+   */
+  private val executors: mutable.HashMap[Int, ExecutorService] =
+    new mutable.HashMap[Int, ExecutorService]()
+
+  /**
+   * Run some code on a particular thread.
+   *
+   * @param threadId Integer id of the thread.
+   * @param task Function to run on the thread.
+   */
+  def run(threadId: Int, task: () => Unit): Unit = {
+    val executor = getOrCreateExecutor(threadId)
+    val future = executor.submit(new Runnable {
+      override def run(): Unit = task()
+    })
+
+    future.get()
+  }
+
+  /**
+   * Delete a particular thread.
+   *
+   * @param threadId Integer id of the thread.
+   */
+  def deleteThread(threadId: Int): Unit = synchronized {
+    executors.remove(threadId).foreach(_.shutdown)
+  }
+
+  /**
+   * Get the executor if it exists, otherwise create a new one.
+   *
+   * @param id Integer id of the thread.
+   * @return The new or existing executor with the given id.
+   */
+  private def getOrCreateExecutor(id: Int): ExecutorService = synchronized {
+    executors.getOrElseUpdate(id, Executors.newSingleThreadExecutor)
```

Comment:
I see the following documentation for `newSingleThreadExecutor`: "Creates an Executor that uses a single worker thread operating off an unbounded queue. (Note however that if this single thread terminates due to a failure during execution prior to shutdown, a new one will take its place if needed to execute subsequent tasks.)"

So, if I do the following on the C# side,

```csharp
try
{
   spark.Read.Parquet("invalidpath"); // should throw on the JVM side.
}
catch (...)
{
  // but I am catching on the C# side
  SparkSession.GetActiveSession(); // Will this be fine?
}
```

Reply:
If `spark.read.parquet()` throws an exception on the JVM side, the executor will not crash because the executor is actually running `DotNetBackendHandler.handleMethodCall()` which has its own try/catch.

Reply:
so there is absolutely no scenario where `newSingleThreadExecutor` will spawn a new thread?

Reply:
Changed the try/catch to catch `Throwable`, so I think we would only lose the executor if some hard crash caused the thread to die.